internal/interfaces/grpc/service_one_port.go

```diff
@@ -87,372 +151,238 @@ func NewServiceOnePort(opts ServiceOptsOnePort) (interfaces.Service, error) {
+}
 
-	withoutUnlockerOnly := false
-	services, err := s.start(withoutUnlockerOnly)
+func (s *serviceOnePort) start(withWalletOnly bool) error {
+	tlsConfig, err := s.opts.tlsConfig()
 	if err != nil {
-		log.WithError(err).Warn(
-			"an error occured while enabling operator and trade interfaces. Shutting down",
-		)
-		panic(nil)
+		return err
+	}
+	server, err := s.newServer(tlsConfig, !withWalletOnly)
+	if err != nil {
+		return err
 	}
 
-	log.Infof("operator, trade interfaces is listening on %s", s.opts.Address)
+	s.server = server
 
-	s.grpcServer = services.grpcServer
-	s.http1Server = services.http1Server
-	s.http2Server = services.http2Server
-	s.mux = services.mux
+	if s.opts.withTls() {
+		go s.server.ListenAndServeTLS("", "")
+	} else {
+		go s.server.ListenAndServe()
+	}
+
+	log.Infof("wallet interface is listening on %s", s.opts.serverAddr())
+	if !withWalletOnly {
+		log.Infof("operator interface is listening on %s", s.opts.serverAddr())
+		log.Infof("trade interface is listening on %s", s.opts.serverAddr())
+	}
+
+	return nil
 }
 
 func (s *serviceOnePort) stop(stopMacaroonSvc bool) {
 	if s.withMacaroons() && stopMacaroonSvc {
 		s.macaroonSvc.Close()
-		log.Debug("stopped macaroon service")
+		log.Debug("closed connection with macaroon db")
 	}
 
-	log.Debug("stop grpc-web server")
-	s.http1Server.Shutdown(context.Background())
-	s.http2Server.Shutdown(context.Background())
-
-	log.Debug("stop grpc server")
-	s.grpcServer.GracefulStop()
-
-	log.Debug("stop mux")
-	s.mux.Close()
+	s.server.Shutdown(context.Background())
+	log.Debug("stopped server")
 }
 
-func (s *serviceOnePort) Stop() {
-	stopMacaroonSvc := true
-	s.stop(stopMacaroonSvc)
-}
+func (s *serviceOnePort) newServer(
+	tlsConfig *tls.Config, withWalletOnly bool,
+) (*http.Server, error) {
+	serverOpts := []grpc.ServerOption{
+		interceptor.UnaryInterceptor(s.macaroonSvc),
+		interceptor.StreamInterceptor(s.macaroonSvc),
+	}
 
-func (s *serviceOnePort) start(withUnlockerOnly bool) (*serviceOnePort, error) {
-	unaryInterceptor := interceptor.UnaryInterceptor(s.macaroonSvc)
-	streamInterceptor := interceptor.StreamInterceptor(s.macaroonSvc)
+	creds := insecure.NewCredentials()
+	if tlsConfig != nil {
+		creds = credentials.NewTLS(tlsConfig)
+	}
+	serverOpts = append(serverOpts, grpc.Creds(creds))
 
 	var adminMacaroonPath string
 	if s.withMacaroons() {
-		adminMacaroonPath = filepath.Join(s.opts.macaroonsDatadir(), AdminMacaroonFile)
+		adminMacaroonPath = filepath.Join(
+			s.opts.macaroonsDatadir(), AdminMacaroonFile,
+		)
 	}
 
-	address := s.opts.Address
-	grpcServer := grpc.NewServer(
-		unaryInterceptor, streamInterceptor,
-	)
-	walletUnlockerHandler := grpchandler.NewWalletUnlockerHandler(
-		s.opts.WalletUnlockerSvc, adminMacaroonPath,
+	grpcServer := grpc.NewServer(serverOpts...)
+	walletHandler := grpchandler.NewWalletHandler(
+		s.opts.AppConfig.WalletService().Wallet(), s.opts.BuildData,
+		adminMacaroonPath,
+		s.onInit, s.onUnlock, s.onLock, s.onChangePwd,
```

Comment:
Doesn't seem so anti-pattern actually https://stackoverflow.com/questions/25203988/go-and-callbacks and looks cleaner for this very specific purpose since we can handle the unlocking of macaroon store in a sync way.